### ISSUE TITLE
problem with implem of mAP@R ?

### ISSUE THREAD

**Issue by drasros** [KevinMusgrave/pytorch-metric-learning#331]
Hi, 

Thanks for open-sourcing this repo!
I want to report a problem which I noticed the the [XBM repo](https://github.com/msight-tech/research-xbm) which uses code similar to yours for mAP@R calculation (I am not sure who's based on who, opening this issue on both). 

I suspect there is a mistake in the implementation of mAP@R:
[Line 99](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/9559b21559ca6fbcb46d2d51d7953166e18f9de6/src/pytorch_metric_learning/utils/accuracy_calculator.py#L99), you divide by max_possible_matches_per_row for this query (which for mAP@R is equal to R for each query).
But what should be done is divide by the __actual__ number of relevant items in the R-best-ranked ones for this query (as is done for [mAP](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/9559b21559ca6fbcb46d2d51d7953166e18f9de6/src/pytorch_metric_learning/utils/accuracy_calculator.py#L97) 2 line later).
wikipedia page on AP: 
![image](https://user-images.githubusercontent.com/16518885/119520716-d27d0300-bd7a-11eb-89c0-a303569d50c3.png)
(by the way, for mAP, as you do line 98, one has to decide what to do with queries which have no relevant items and on which AP is therefore not defined. You use max_possible_values_per_row=0 for these rows, but I am not sure this is the most standard choice, maybe it's good to give to option to set AP=0, or AP=1, or exclude these rows...)

Regards,
A

**Comment by drasros** [comment]
(To confirm this problem I ran some test of the xbm implem vs sklearn implem. Your implem is slightly different so I don't 100% claim that it is wrong, just suspect it and want to kindly report the suspiscion )

**Comment by KevinMusgrave** [comment]
> I want to report a problem which I noticed the the [XBM repo](https://github.com/msight-tech/research-xbm) which uses code similar to yours for mAP@R calculation (I am not sure who's based on who, opening this issue on both).

The AccuracyCalculator code originates from this repo (pytorch-metric-learning)

> I suspect there is a mistake in the implementation of mAP@R:
> [Line 99](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/9559b21559ca6fbcb46d2d51d7953166e18f9de6/src/pytorch_metric_learning/utils/accuracy_calculator.py#L99), you divide by max_possible_matches_per_row for this query (which for mAP@R is equal to R for each query).
> But what should be done is divide by the **actual** number of relevant items in the R-best-ranked ones for this query (as is done for [mAP](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/9559b21559ca6fbcb46d2d51d7953166e18f9de6/src/pytorch_metric_learning/utils/accuracy_calculator.py#L97) 2 line later).

I think the implementation is correct. Here is the definition I'm using

![image](https://user-images.githubusercontent.com/2314891/119554504-bd987380-bd6a-11eb-8002-93899a3a100a.png)

The divisor is different from the usual mAP. I probably could have chosen a better name than "mAP@R".

> (by the way, for mAP, as you do line 98, one has to decide what to do with queries which have no relevant items and on which AP is therefore not defined. You use max_possible_values_per_row=0 for these rows, but I am not sure this is the most standard choice, maybe it's good to give to option to set AP=0, or AP=1, or exclude these rows...)

Good point. Actually if you're using AccuracyCalculator (and not the ```mean_average_precision``` function directly) then the queries with no relevant items should be filtered out in this function: https://github.com/KevinMusgrave/pytorch-metric-learning/blob/9559b21559ca6fbcb46d2d51d7953166e18f9de6/src/pytorch_metric_learning/utils/accuracy_calculator.py#L185-L191

which is called here:
https://github.com/KevinMusgrave/pytorch-metric-learning/blob/9559b21559ca6fbcb46d2d51d7953166e18f9de6/src/pytorch_metric_learning/utils/accuracy_calculator.py#L308-L328

**Comment by drasros** [comment]
> The divisor is different from the usual mAP. I probably could have chosen a better name than "mAP@R".

Thanks for your reply. I see. Yes, the name is somewhat misleading in the sense that this metric is not the mean of AP over the subset of R top-ranked candidates, but rather the full-mAP where the sum is truncated to relevants in top-R. 

(the same confusion can arise for mAP@k, what do we do the integral on?.. I am not sure what the most standard thing is. )
